### PR TITLE
Flatten modules into main graph

### LINKED ISSUE #1582: Destroy doesn't respect module dependencies correctly
I've ran into an issue similar to #1472 and I think i've distilled it to a simple example.

This is on the current master of terraform:
```
$ terraform version
Terraform v0.5.0-dev (d9e8c112aec46ff62805833ced449364e3115290)
```

Given a configuration:

**./main.tf**
```
module "a_module" {
  source = "./a_module"
}

resource "null_resource" "b" {
  blah = "${module.a_module.a_output}"
}
```

**./a_module/main.tf**
```
resource "null_resource" "a" {
}

output "a_output" {
  value = "${null_resource.a.id}"
}
 ```

On apply, it should bring up `a_module.null_resource.a` before it brings up `root.null_resource.b` and on destroy, it should destroy `b` before `a` by respecting the dependency. 

So on apply I get:

```
$ terraform apply
module.a_module.null_resource.a: Creating...
module.a_module.null_resource.a: Creation complete
null_resource.b: Creating...
null_resource.b: Creation complete

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
use the `terraform show` command.

State path: terraform.tfstate
```

Good there..

In the `terraform.tfstate`, I have:
```json
{
    "version": 1,
    "serial": 1,
    "modules": [
        {
            "path": [
                "root"
            ],
            "outputs": {},
            "resources": {
                "null_resource.b": {
                    "type": "null_resource",
                    "depends_on": [
                        "module.a_module"
                    ],
                    "primary": {
                        "id": "4686510281091019717",
                        "attributes": {
                            "id": "4686510281091019717"
                        }
                    }
                }
            }
        },
        {
            "path": [
                "root",
                "a_module"
            ],
            "outputs": {
                "a_output": "7727900350613538303"
            },
            "resources": {
                "null_resource.a": {
                    "type": "null_resource",
                    "primary": {
                        "id": "7727900350613538303",
                        "attributes": {
                            "id": "7727900350613538303"
                        }
                    }
                }
            }
        }
    ]
}
```

But when I destroy:
```
$ terraform destroy -force
module.a_module.null_resource.a: Refreshing state... (ID: 7727900350613538303)
null_resource.b: Refreshing state... (ID: 4686510281091019717)
null_resource.a: Destroying...
null_resource.a: Destruction complete
null_resource.b: Destroying...
null_resource.b: Destruction complete

Apply complete! Resources: 0 added, 0 changed, 2 destroyed.
```

It's destroying `a` before `b`, which I wouldn't expect. This has led to problems in which a module that we depend on creates something which we add to it in our own config. But when we destroy, the service behind the "parent" blows up because it's not empty. If destroyed the leaves first, it wouldn't have this problem.

### PR DESCRIPTION
**Pull request by mitchellh**
Fixes #1582 

Might affect #1637 (hopefully fixes, unsure due to private repo). 

All context tests pass but this is a pretty big change and I'll want to try at least on our core ops repo before merging (since we use modules heavily). 

This changes the way the graph treats modules. Prior to this PR, modules were their own isolated subgraphs. When evaluating, the entire subgraph would get walked in one step. The problem with this is that it turns out that modules can't always be seen as a single executable unit. 

This PR flattens the subgraphs of modules into the main root graph, then does the destroy transformations _after_ this flattening. This lets the destroy edge manipulations have a global view of the graph, and create the proper ordering of things.

Most of this was surprisingly straightforward. The only part I'm really unhappy about is the special casing in graph builder to only do some steps for the root path. 

I think a future change/refactor can probably simplify this even further by just getting rid of static subgraphs altogether and just having the transform add all the nodes for all the modules at once. I was about to go down this route, but decided that if this works, we should get this in first, and refactor later, especially given that the changes here aren't hideous. 

I have a feeling this will fix weird module-boundary create-before-destroy issues as well but I don't have a test case yet to reproduce anything of that sort so I can't confirm.

Additionally, @armon raised good points about this will hurt performance of Terraform for very large graphs. This is because the O(N) lookups for graph vertices will get bad since N Is getting quite large (before it was N within each subgraph). Given that almost all O(N) lookups are for interface implementations, I think we can just do interface-specific lookaside tables in the future and speed everything up to O(1).

@phinze: One thing for you is to ask if the graph functionality looks good? I don't want to affect the `terraform graph` output we did for 0.5 for this.

### CODE REVIEW

File: terraform/context_test.go [hunk 1]
```diff
@@ -4053,6 +4053,88 @@ func TestContext2Apply_module(t *testing.T) {
...
+
+	{
+		actual := strings.TrimSpace(state.String())
+		expected := strings.TrimSpace(testTerraformApplyModuleDestroyOrderStr)
+		if actual != expected {
+			t.Fatalf("bad: \n%s", actual)
+		}
+	}
+}
+
```
Comment (phinze) [inline terraform/context_test.go]: I love how the big overarching test comes first in the diff. :smile: 

_"WE GON' MAKE THIS WORK!"_

File: terraform/eval_context_builtin.go [hunk 2]
```diff
@@ -14,6 +14,8 @@ import (
 type BuiltinEvalContext struct {
 	PathValue           []string
 	Interpolater        *Interpolater
+	InterpolaterVars    map[string]map[string]string
```
Comment (phinze) [inline terraform/eval_context_builtin.go]: I'd love a quick high level explanation for the Interp vars stuff. I can see the pieces but it'd be helpful to hear the overall story.

File: terraform/graph_builder.go [hunk 3]
```diff
@@ -100,7 +100,7 @@ func (b *BuiltinGraphBuilder) Build(path []string) (*Graph, error) {
 
 // Steps returns the ordered list of GraphTransformers that must be executed
 // to build a complete graph.
-func (b *BuiltinGraphBuilder) Steps() []GraphTransformer {
+func (b *BuiltinGraphBuilder) Steps(path []string) []GraphTransformer {
```
Comment (phinze) [inline terraform/graph_builder.go]: I needed this too for module targeting. I'll rebase on top of this.

File: terraform/graph_builder.go [hunk 4]
```diff
@@ -134,24 +134,41 @@ func (b *BuiltinGraphBuilder) Steps() []GraphTransformer {
...
+
+			// Make sure we have a single root after the above changes.
+			// This is the 2nd root transformer. In practice this shouldn't
+			// actually matter as the RootTransformer is idempotent.
+			&RootTransformer{},
+
+			// Perform the transitive reduction to make our graph a bit
+			// more sane if possible (it usually is possible).
+			&TransitiveReductionTransformer{},
+		)
```
Comment (phinze) [inline terraform/graph_builder.go]: Ah yeah this is funky, but agreed with top-level description that it's better to land it working and then refactor.

File: terraform/graph_config_node.go [hunk 5]
```diff
@@ -45,634 +39,3 @@ type GraphNodeTargetable interface {
...
-				},
-			},
-		},
-	}
-}
-
-// GraphNodeSubgraph impl.
-func (n *graphNodeModuleExpanded) Subgraph() *Graph {
-	return n.Graph
-}
```
Comment (phinze) [inline terraform/graph_config_node.go]: With the file split it's sorta hard to read for differences in the graph nodes - can you highlight the important areas where the node code changes for me? <3

Comment (phinze) [inline terraform/graph_config_node.go]: Actually n.m. after a bunch of scrolling I think I get it - just `GraphNodeFlattenable` and `GraphNodeProxy` impls and a few helpers.

File: terraform/graph_config_node_module.go [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+// GraphNodeFlattenable impl.
+func (n *graphNodeModuleExpanded) FlattenGraph() *Graph {
+	graph := n.Subgraph()
+	input := n.Original.Module.RawConfig
+
+	// Go over each vertex in the graph and wrap the configuration
+	// items so that the dependencies properly map to the modules.
+	// See the docs for graphNodeModuleWrappable for more info.
```
Comment (phinze) [inline terraform/graph_config_node_module.go]: is `graphNodeModuleWrappable` from an earlier iteration? do you mean `graphNodeModuleSkippable` now?

Comment (mitchellh) [inline terraform/graph_config_node_module.go]: Oh yeah, it is, I'll update the comment.

File: terraform/eval_context_builtin.go (hunk 2 above)
Comment (mitchellh) [inline terraform/eval_context_builtin.go]: I just added a comment on the new diff.

**Comment by phinze** [comment]
Running this on a local sample infra with a cycle I _thought_ would have been fixed by this yielded unexpected results. Digging in and I'll follow up.

**Comment by ketzacoatl** [comment]
@mitchellh I can get you access to the repo if you would like to test with it in some capacity. I can also test that repo with this updated revision. That repo will be public soon..

**Comment by mitchellh** [comment]
@ketzacoatl I worked with @phinze and verified we've fixed your bvug.

**Comment by ketzacoatl** [comment]
very exciting!

**Pull request merged**